### ISSUE TITLE
Quest  :  Add MLS Support for App configs 

### ISSUE THREAD

**Issue by f-odhiambo** [opensrp/fhircore#987]
**Name of feature to enhance**
Add multilanguage support for App configs in quest 

**Description of feature**
Add the ability to alternative translated  config string based on the language that has been selected or retrieved in shared preference 

**Describe the enhancement**
Add ability to have configs in multiple language translates and ability to select/read them based on the language preference at runtime 

**Sample Config** 
```
{
  "appId": "quest",
  "classification": "patient_register",
  "appTitle": "Clients",
  "filterText": "Show overdue",
  "searchBarHint": "Search for ID or client name",
  "newClientButtonText": "Register new client",
  "newClientButtonStyle": "rounded_corner",
  "showSearchBar": true,
  "showFilter": false,
  "showScanQRCode": false,
  "showNewClientButton": true,
  "showSideMenu": false,
  "showBottomMenu": true,
  "registrationForm": "3435",
  "primaryFilter": {
    "key": "_tag",
    "system": "http://fhir.ona.io",
    "code": "000003"
  }
}
```

**Translated Config snippet**  

`searchBarHint: ["en": "Search your name", "sw": "Seeerchhh ur Neem"]`

**Comment by f-odhiambo** [comment]
@ndegwamartin @ellykits Let me know if this is feasible or we can have a different

**Comment by f-odhiambo** [comment]
This also affects some portions of the app such as the once highlighted in the below screenshots
1. Test Results 
2. Take photo
3. Control mode intro instructions 
4. Any other item that is in a cogi will require some translation 

**Screenshots**
<img src="https://user-images.githubusercontent.com/4540684/150491260-12eb6902-f68b-4751-b55a-05c409681354.png" width="200" height="400" />
<img src="https://user-images.githubusercontent.com/4540684/150491271-5d4438a2-d7ee-4e08-af1d-8276908059ee.png" width="200" height="400" />
<img src="https://user-images.githubusercontent.com/4540684/150491276-50f2d115-213d-46c4-8c58-177b0623e616.png" width="200" height="400" />
<img src="https://user-images.githubusercontent.com/4540684/150491281-74c7f11f-46c7-41d9-91c4-0142711dda2a.png" width="200" height="400" />
<img src="https://user-images.githubusercontent.com/4540684/150491285-0f0ec0d0-8e7c-4b65-bee2-9bd070db69d4.png" width="200" height="400" />

**Comment by ndegwamartin** [comment]
This is a good suggestion, however I'd recommend the content inside to be valid json so that processing is made easy with the available tools/utilities.

Perhaps we can try borrow from FHIR where they the concept of extensions and also have individual array object for the different translations. We could have for the above e.g. 

```
  "searchBarHint": "Search for ID or client name",
   "_searchBarHint": {
    "translations": [
      {
        "language": "sw",
        "value": "Tafuta na ID au jina"
      },
      {
        "language": "fr",
        "value": "Recherche par ID ou nom"
      }
    ]
  }
```

**Comment by ndegwamartin** [comment]
Here's the FHIR documentation where these are borrowed from [HL7 FHIR i18n](http://hl7.org/fhir/languages.html##ext)

**Comment by pld** [comment]
could we do it in the value for the specific element:

```json
"searchBarHint": [
      {
        "language": "sw",
        "value": "Search for ID or client name"
      },
      {
        "language": "sw",
        "value": "Tafuta na ID au jina"
      },
      {
        "language": "fr",
        "value": "Recherche par ID ou nom"
      }
]
```

**Comment by ndegwamartin** [comment]
I suppose then for that approach we have to introduce an new field in the parent json that represents the _default locale_ for the configuration, so that we know which translation to select from the array list. e.g.

```
 { 
    "language" : "en",
    "searchBarHint": [
          {
            "language": "en",
            "value": "Search for ID or client name"
          },
          {
            "language": "sw",
            "value": "Tafuta na ID au jina"
          }
```
  
  
  Which adds the template here `"language" : "<code>" // Default language of the configuration content,`

**Comment by pld** [comment]
Why do we need the default locale? I was thinking the current locale in the app would be sufficient

**Comment by ndegwamartin** [comment]
Oh I see. Yeah if we pick the locale then we needn't worry about a default locale for the config.

I do suppose we do need though a resolution policy for when the default locale does not have a translation present in the array. Maybe something like pick the first or last item.

**Comment by pld** [comment]
Yea true, we could have a default config that has a default locale

**Comment by ndegwamartin** [comment]
Can we then go with the configuration format here? 

```json

  "searchBarHint": "Search for ID or client name",
  "_searchBarHint": [
    {
      "language": "sw",
      "value": "Tafuta na ID au jina"
    },
    {
      "language": "fr",
      "value": "Recherche par ID ou nom"
    }
  ]
  
```
This has a default value as well as any available translations

**Comment by pld** [comment]
What I meant was use the format in https://github.com/opensrp/fhircore/issues/987#issuecomment-1021351137, but that there's a separate config, like for an app, that has lines in it like

```json
...
"default_locale": "en"
...
"available_locales": ["en", "sw"]
...
```

Telling it what the default locale is and what locales should be made available, like if the app supports French, but it's in Kenya, we may want to not show French as an option.

We'll need a hard-coded version of these config keys, so that if a user doesn't set it takes those.

**Comment by ellykits** [comment]
I would recommend we use this format suggested by @pld 

```json
"searchBarHint": [
      {
        "language": "en",
        "value": "Search for ID or client name"
      },
      {
        "language": "sw",
        "value": "Tafuta na ID au jina"
      },
      {
        "language": "fr",
        "value": "Recherche par ID ou nom"
      }
]
```

The current configurations will be refactored into [HL7 Parameters](https://www.hl7.org/fhir/parameters.html) resource. The format above will ease the transformation. We can set the default locale and available locales configs in the the application configuration in the composition resource.

**Comment by ndegwamartin** [comment]
@ellykits based on our discussion on using [HL7 Parameters resource](https://www.hl7.org/fhir/parameters.html) for configuration, might it be better then to choose a format that exactly resembles the final one? 

That way, only the source of configuration (instead of the logic) will need to change or be refactored

**Comment by ellykits** [comment]
@dubdabasoduba

**Comment by dubdabasoduba** [comment]
> **Name of feature to enhance** Add multilanguage support for App configs in quest
> 
> **Description of feature** Add the ability to alternative translated config string based on the language that has been selected or retrieved in shared preference
> 
> **Describe the enhancement** Add ability to have configs in multiple language translates and ability to select/read them based on the language preference at runtime
> 
> **Sample Config**
> 
> ```
> {
>   "appId": "quest",
>   "classification": "patient_register",
>   "appTitle": "Clients",
>   "filterText": "Show overdue",
>   "searchBarHint": "Search for ID or client name",
>   "newClientButtonText": "Register new client",
>   "newClientButtonStyle": "rounded_corner",
>   "showSearchBar": true,
>   "showFilter": false,
>   "showScanQRCode": false,
>   "showNewClientButton": true,
>   "showSideMenu": false,
>   "showBottomMenu": true,
>   "registrationForm": "3435",
>   "primaryFilter": {
>     "key": "_tag",
>     "system": "http://fhir.ona.io",
>     "code": "000003"
>   }
> }
> ```
> 
> **Translated Config snippet**
> 
> `searchBarHint: ["en": "Search your name", "sw": "Seeerchhh ur Neem"]`

These configs are going to change and use Parameters after configurable app work is complete we also need to think about that.

**Comment by ndegwamartin** [comment]
> @ellykits based on our discussion on using [HL7 Parameters resource](https://www.hl7.org/fhir/parameters.html) for configuration, might it be better then to choose a format that exactly resembles the final one?
> 
> That way, only the source of configuration (instead of the logic) will need to change or be refactored

@dubdabasoduba I think your input was required here. That is defining the final format for the MLS support using [HL7 Parameters resource](https://www.hl7.org/fhir/parameters.html)

**Comment by dubdabasoduba** [comment]
> 

@ndegwamartin thanks for the clarification let me look at it.

**Comment by dubdabasoduba** [comment]
> > @ellykits based on our discussion on using [HL7 Parameters resource](https://www.hl7.org/fhir/parameters.html) for configuration, might it be better then to choose a format that exactly resembles the final one?
> > That way, only the source of configuration (instead of the logic) will need to change or be refactored
> 
> @dubdabasoduba I think your input was required here. That is defining the final format for the MLS support using [HL7 Parameters resource](https://www.hl7.org/fhir/parameters.html)

@ndegwamartin thanks for the clarification let me look at it.

**Comment by dubdabasoduba** [comment]
@ndegwamartin I would say let's combine these 2 https://github.com/opensrp/fhircore/issues/987#issuecomment-1024797981 & https://github.com/opensrp/fhircore/issues/987#issuecomment-1024797981. That we abstract the locale definitions and make sure we can even use them in other places.

**Comment by ndegwamartin** [comment]
@dubdabasoduba I see, so something like:

```json
"default_locale": "en",
"searchBarHint": [
      {
        "language": "en",
        "value": "Search for ID or client name"
      },
      {
        "language": "sw",
        "value": "Tafuta na ID au jina"
      },
      {
        "language": "fr",
        "value": "Recherche par ID ou nom"
      }
]
```

**Comment by dubdabasoduba** [comment]
@ndegwamartin If you look at this `config_application.json` we have the languages added there. Maybe we can add the default_locale there too. 

The other format below is okay by me. 
```json
"searchBarHint": [
      {
        "language": "en",
        "value": "Search for ID or client name"
      },
      {
        "language": "sw",
        "value": "Tafuta na ID au jina"
      },
      {
        "language": "fr",
        "value": "Recherche par ID ou nom"
      }
]
````

**Comment by pld** [comment]
Yea that's correct, the language to be used will not be defined by the resource, but by an external configuration.

**Comment by dubdabasoduba** [comment]
> Yea that's correct, the language to be used will not be defined by the resource, but by an external configuration.

Agreed. We should have the same language support all through the app. That way it's easier to maintain context even in code.

**Comment by ndegwamartin** [comment]
Ok. Assuming that that the resource has lets say a lot of fields, we probably need a way to to efficiently identify:
1. This is a translation content json field
2. This translation content field is for this particular field

**Comment by ndegwamartin** [comment]
> Can we then go with the configuration format here?
> 
> ```json
>   "searchBarHint": "Search for ID or client name",
>   "_searchBarHint": [
>     {
>       "language": "sw",
>       "value": "Tafuta na ID au jina"
>     },
>     {
>       "language": "fr",
>       "value": "Recherche par ID ou nom"
>     }
>   ]
>   
> ```
> 
> This has a default value as well as any available translations

The above might bring us back to [this format here](https://github.com/opensrp/fhircore/issues/987#issuecomment-1072044264)

**Comment by ndegwamartin** [comment]
> Can we then go with the configuration format here?
> 
> ```json
>   "searchBarHint": "Search for ID or client name",
>   "_searchBarHint": [
>     {
>       "language": "sw",
>       "value": "Tafuta na ID au jina"
>     },
>     {
>       "language": "fr",
>       "value": "Recherche par ID ou nom"
>     }
>   ]
>   
> ```
> 
> This has a default value as well as any available translations

[The above](https://github.com/opensrp/fhircore/issues/987#issuecomment-1072044264) might bring us back to this format here.

**Comment by dubdabasoduba** [comment]
@ndegwamartin 
My thinking was the `config_application.json` file will just describe what the default language is. The format below will be used for all the translations. That means if `en` is the default it will be picked from the JSONArray. This way we don't have to add an extra attribute to the other languages. We use the same attribute we just add a new JSONObject to the JSONArray. 

``` json
"searchBarHint": [
      {
        "language": "en",
        "value": "Search for ID or client name"
      },
      {
        "language": "sw",
        "value": "Tafuta na ID au jina"
      },
      {
        "language": "fr",
        "value": "Recherche par ID ou nom"
      }
]
```

**Comment by ndegwamartin** [comment]
Transifex's structure for translatable strings looks a bit different, see https://docs.transifex.com/formats/json/structured-json. 
Not sure it supports ability to integrate custom formats, otherwise we'd have to conform to the above standard.

**Comment by pld** [comment]
Not sure how this converts, would following this format look like:

```json
"searchBarHint": {
      "en": {
        "string": "Search for ID or client name"
      },
      "sw": {
        "string": "Tafuta na ID au jina"
      },
      "fr": {
        "string": "Recherche par ID ou nom"
      }
]
```

that seems fine

**Comment by ndegwamartin** [comment]
Actually I think it would be more like:
 
**EN** file
```json
 {
      "searchBarHint": {
        "string": "Search for ID or client name"
      }
 }
```

**SW** file:
```json
 {
      "searchBarHint": {
        "string": "Tafuta na ID au jina"
      }
 }
```
**FR** file
```json
 {
      "searchBarHint": {
        "string": "Recherche par ID ou nom"
      }
 }
```

**Comment by dubdabasoduba** [comment]
> ah ok cool, yea that seems fine too, what do you think?

I think it's fine but it means a whole different file for each language translation.  I don't think that's the route we want to follow.  I think I would rather we use the format highlighted below as we try to figure out how we can use property files as we did on NativeForms. This way we maintain 1 JSON file and the multiple `.properties` to the languages. 

```json
"searchBarHint": [
      {
        "language": "en",
        "value": "Search for ID or client name"
      },
      {
        "language": "sw",
        "value": "Tafuta na ID au jina"
      },
      {
        "language": "fr",
        "value": "Recherche par ID ou nom"
      }
]
```

**Comment by ndegwamartin** [comment]
Based on previous discussions, here's the proposed implementation plan we went with:

- Have the app json configs with a placeholder syntax
- Sync the different language translations as `.properties` Binary Resources
- Persist them on local and use them via resource bundles or interpolation

**Sample.json config file:**

```json
{
   "appId": "quest",
   "classification": "login",
   "applicationName": "{{ application.name }}",
   "applicationVersion": "0.0.2",
   "darkMode": false,
   "showLogo": false,
   "enablePin": false
 }
```

**Sample language translation files:**

_Sample.json.properties_en_
`application.name=Hello Quest`

_Sample.json.properties_sw_
`application.name=Habari Quest`

**Comment by FikriMilano** [comment]
@ndegwamartin @pld 
I'm thinking of using nested `section` for this case
https://build.fhir.org/composition.html
```
{
  "title": "App configuration",
  "focus": {
    "reference": "Binary/128867",
    "identifier": {
      "value": "application"
    }
  },
  "section": [
    {
      "title": "Translation English",
      "focus": {
        "reference": "Binary/xxxxx",
        "identifier": {
          "value": "application_en"
        }
      }
    },
    {
      "title": "Translation Swahili",
      "focus": {
        "reference": "Binary/xxxxx",
        "identifier": {
          "value": "application_sw"
        }
      }
    }
  ]
}
```

**Comment by f-odhiambo** [comment]
LGTM @ndegwamartin

**Comment by ndegwamartin** [comment]
To be tracked/closed as part of #1455
